sled-agent/src/bootstrap/agent.rs

```diff
@@ -729,6 +746,163 @@ impl Agent {
+        Ok(())
+    }
+
+    async fn uninstall_sled_local_config_locked(
+        &self,
+        _state: &tokio::sync::MutexGuard<'_, SledAgentState>,
+    ) -> Result<(), BootstrapError> {
+        tokio::fs::remove_dir_all(omicron_common::OMICRON_CONFIG_PATH)
+            .await
+            .or_else(|err| match err.kind() {
+                std::io::ErrorKind::NotFound => Ok(()),
+                _ => Err(err),
+            })
+            .map_err(|err| BootstrapError::Io {
+                message: format!(
+                    "Deleting {}",
+                    omicron_common::OMICRON_CONFIG_PATH
+                ),
+                err,
+            })?;
+        tokio::fs::create_dir_all(omicron_common::OMICRON_CONFIG_PATH)
+            .await
+            .map_err(|err| BootstrapError::Io {
+                message: format!(
+                    "Creating config directory {}",
+                    omicron_common::OMICRON_CONFIG_PATH
+                ),
+                err,
+            })?;
+        Ok(())
+    }
+
+    async fn uninstall_networking_locked(
+        &self,
+        _state: &tokio::sync::MutexGuard<'_, SledAgentState>,
+    ) -> Result<(), BootstrapError> {
+        // NOTE: This is very similar to the invocations
+        // in "sled_hardware::cleanup::cleanup_networking_resources",
+        // with a few notable differences:
+        //
+        // - We can't remove bootstrap-related networking -- this operation
+        // is performed via a request on the bootstrap network.
+        // - We avoid deleting addresses using the chelsio link. Removing
+        // these addresses would delete "cxgbe0/ll", and could render
+        // the sled inaccessible via a local interface.
+
+        sled_hardware::cleanup::delete_underlay_addresses(&self.log)
+            .map_err(BootstrapError::Cleanup)?;
+        sled_hardware::cleanup::delete_omicron_vnics(&self.log)
+            .await
+            .map_err(BootstrapError::Cleanup)?;
+        illumos_utils::opte::delete_all_xde_devices(&self.log)?;
+        Ok(())
+    }
+
+    async fn uninstall_storage_locked(
+        &self,
+        _state: &tokio::sync::MutexGuard<'_, SledAgentState>,
+    ) -> Result<(), BootstrapError> {
+        let datasets = zfs::get_all_omicron_datasets()
+            .map_err(BootstrapError::ZfsDatasetsList)?;
+        for dataset in &datasets {
+            info!(self.log, "Removing dataset: {dataset}");
+            zfs::Zfs::destroy_dataset(dataset)?;
+        }
+
+        Ok(())
+    }
+
+    /// Resets this sled, removing:
+    ///
+    /// - All control plane zones (except the switch zone)
+    /// - All sled-local configuration
+    /// - All underlay networking
+    /// - All storage managed by the control plane
+    ///
+    /// This API is intended to put the sled into a state where it can
+    /// subsequently be initialized via RSS.
+    pub async fn sled_reset(&self) -> Result<(), BootstrapError> {
```

Comment:
This function is worth reviewing; it's the implementation of all "local-to-my-sled" resetting